components/autofill/src/sync/address/outgoing.rs

```diff
@@ -256,4 +293,46 @@ mod tests {
             COLLECTION_NAME.into(),
         );
     }
+
+    #[test]
+    fn test_outgoing_roundtrip_unknown() {
+        let mut db = new_syncable_mem_db();
+        let tx = db.transaction().expect("should get tx");
+        let ao = OutgoingAddressesImpl {};
+
+        // create synced record with non-zero sync_change_counter
+        let mut test_record = test_record('D');
+        let initial_change_counter_val = 2;
+        test_record.metadata.sync_change_counter = initial_change_counter_val;
+        assert!(add_internal_address(&tx, &test_record).is_ok());
+        // put "unknown_fields" into the mirror payload to imitate the server
+        let unknown_fields: UnknownFields =
+            serde_json::from_value(json! {{ "foo": "bar", "baz": "qux"}}).unwrap();
+        test_insert_mirror_record(&tx, test_record.clone(), unknown_fields);
+        exists_with_counter_value_in_table(
+            &tx,
+            DATA_TABLE_NAME,
+            &test_record.guid,
+            initial_change_counter_val,
+        );
+
+        let outgoing = &ao
+            .fetch_outgoing_records(&tx, COLLECTION_NAME.into())
+            .unwrap();
+        // Ensure we have our unknown values for the roundtrip
+        let bso_payload: Map<String, Value> =
+            serde_json::from_str(&outgoing.changes[0].payload).unwrap();
+        let entry = bso_payload.get("entry").unwrap();
+        assert_eq!(entry.get("foo").unwrap(), "bar");
+        assert_eq!(entry.get("baz").unwrap(), "qux");
+        do_test_outgoing_synced_with_local_change(
+            &tx,
+            &ao,
+            &test_record.guid,
+            DATA_TABLE_NAME,
+            MIRROR_TABLE_NAME,
+            STAGING_TABLE_NAME,
+            COLLECTION_NAME.into(),
+        );
+    }
```

Comment:
How about some more tests?

- **3 way merge:** there's a record in the mirror table with `unknown_fields` set, an incoming record with new `unknown_fields`, and local changes.  I think we should take the new incoming `unknown_fields` in this case.
- **Fork:** Like before there's a mirror record with `unknown_fields`, an incoming record with changes, and a local record with changes.  In this case, the incoming and local record have conflicts, so we "fork" the record.  I'm not totally sure how this should work, but I think:
  - The forked local record should have the `unknown_fields` from the existing mirror record.  There should be a new mirror record with corresponding values.
  - We shouldn't include the incoming record in the outgoing changes -- this means it will keep it's `unknown_fields` values.  We should update the mirror record to match the incoming record.

Reply:
Thanks for the review! Hmmm, definitely something to think about more (and maybe add appropriate tests to reflect what's actually happening). However, if I'm thinking of this right...

1. The three-way merge is happening on the `InternalAddress` struct, which `unknown_fields` is not a part of -- due to `unknown_fields` not being in the local db and strictly on the mirror table
2. Upon outgoing (so I presume all incoming issues have been merged/planned) we take the the `unknown_fields` from the `mirror` table, which means it should be whatever was latest fetched from the server

So I'm thinking that `unknown_fields` should *never* be in a 3-way merge situation, at least with the current code. 

Forking is a tad bit more complicated, because according to `apply_incoming_action` https://github.com/mozilla/application-services/blob/b14fd3566879f927901691407c0341061bbdeccb/components/autofill/src/sync/mod.rs#L370-L374

we change the local guid, so when we try to get the corresponding `unknown_fields` from the mirror, it'll of course not be there. There is also not an easy way to keep track it's previous guid

Reply:
Ben's note that we need tests there is excellent.

Re 3-way merge, I agree the mirror should end up with whatever unknown fields were incoming, and I think the current impl will correctly do that, but should have a test which proves it.

Re forking: ugh. I think it should actually be possible to just change the guid in the mirror table too, so long as we make sure the record has an appropriate change counter (which I think it will)? Again, test demonstrating that is important.

Reply:
Changing the GUID on the mirror table seems scary to me.  How will other clients see that change when they sync next?

Could `change_local_guid()` also update the GUID in the mirror table so that the forked record would see that when it came to upload the changes, or is it too late by then since the new mirror record was already added.

Reply:
> Changing the GUID on the mirror table seems scary to me. How will other clients see that change when they sync next?

> Could `change_local_guid()` also update the GUID in the mirror table so that the forked record would see that when it came to upload the changes, or is it too late by then since the new mirror record was already added.

I'm slightly confused here - that second sentence is exactly what I meant, but the first sentence says you think that is scary?

Reply:
I wonder if I'm thinking about the forking right -- from what I understand looking at the code:

1. We have a mirror record that has `given-name` = "mir"
2. We have a record that changed locally `given-name` = "loc"
3. We have a incoming record that also changed it `given-name` = "inc"

Given that it's impossible to reconcile this -- we implement a forking strategy, which means
1. The mirror record now has `inc`
2. The local record now has `inc`
3. A NEW local record has `loc` <--> This does not exist on the server yet and will be written as a new record on next sync


So the current suggestion is to change the `guid` of `1` to match the guid of `3` --> Which I think may be a bad idea as the data doesn't match given that the "server wins" for the existing record. A good visualization of this is in the `test_reconcile.rs` json (which is where I'm adding my additional testing).

https://github.com/mozilla/application-services/blob/d8cdb7a15ff27a25336265e49a1936147789cd7e/components/autofill/src/sync/tests/test_reconcile.rs#L279-L310

Reply:
Ah yeah, I think you are correct - at the time we detect we need to fork, we've already overwritten the mirror record with any new incoming fields - we've then lost what the unknown fields were.

So I'm on-board with that - the forked record gets no unknown fields - in effect, it looks like a brand-new record created from scratch on this device, which seems perfectly fine to me.

Reply:
> Ah yeah, I think you are correct - at the time we detect we need to fork, we've already overwritten the mirror record with any new incoming fields

Actually, I don't think that's true - we've *staged* the incoming record in `addresses_sync_staging` - `addresses_mirror` should have the old version until `finish_incoming()` is called?

Reply:
> > Ah yeah, I think you are correct - at the time we detect we need to fork, we've already overwritten the mirror record with any new incoming fields
> 
> Actually, I don't think that's true - we've _staged_ the incoming record in `addresses_sync_staging` - `addresses_mirror` should have the old version until `finish_incoming()` is called?

That's correct, this turned out to be somewhat easy and I added it as a separate commit to see the changes for this. I still think there is a theoretical question here of _do we want the forked record to have an accompanying local mirror_ -- it'll get uploaded all the same and treated as a "changed" record rather than "new".